File: Codeforces/1713/1713E-dsu.py
###
# Solution: Disjoint Set Union (DSU), TLE with python, but AC with C++
# we can find change k twice equals to no change, so for each k, we have two states: change k once (co) or nochange (nc)
# for any pair A_ij and A_ji, we only concern i != j condition, then
# we change the pair <==> i and j have the different state (co v.s. nc);
# we unchange the pair <==> i and j have the same state
# we can change or unchange the pair <==> i and j have no constraint
# for the lexicographically smallest matrix, we need to concern any (i, j) pair (i > j) from the i * n + j priority (from small to large)
# thus we can use this criteria to build a DSU, also to store the stage: whether this node's state equals to its parent node
###

def get_parent(union_set, node):
    if union_set[node][0] == node:
        return union_set[node]
    else:
        parent, flag = union_set[node]
        root, flag_p = get_parent(union_set, parent)
        flag_new = (flag + flag_p) % 2
        union_set[node] = [root, flag_new]
        return union_set[node]

def merge(union_set, nodeA, nodeB, flag):
    parA, flagA = get_parent(union_set, nodeA)
    parB, flagB = get_parent(union_set, nodeB)
    if parA != parB:
        union_set[parB] = [parA, (flag+flagA+flagB)%2]

t = int(input())
for _ in range(t):
    n = int(input())
    matrix = []
    for i in range(n):
        arr = input().split()
        arr = [int(__) for __ in arr]
        matrix.append(arr)

    union_set = [[i, 0] for i in range(n)] # 0: same with root, 1: unsame with root
    for i in range(n):
        for j in range(i+1, n):
            if matrix[i][j] < matrix[j][i]:
                merge(union_set, i, j, 0)
            elif matrix[i][j] > matrix[j][i]:
                merge(union_set, i, j, 1)
            else:
                pass

    for i in range(n):
        __, flag = get_parent(union_set, i)
        if flag == 1: # change for all flag == 1
            for j in range(n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

    for i in range(n):
        print(' '.join([str(__) for __ in matrix[i]]))

File: Codeforces/1713/1713E-dsu.cpp
/*
See solution in the code file 1713E.py
*/

#include <iostream>
using namespace std;

void get_parent(int* union_set, int* flags_set, int node) {
    if (union_set[node] == node) {
        return;
    } else {
        int parent = union_set[node];
        int flag = flags_set[node];

        get_parent(union_set, flags_set, parent);
        int root = union_set[parent];
        int flag_root = flags_set[parent];

        union_set[node] = root;
        flags_set[node] = (flag + flag_root) % 2;
    }
}

bool merge(int* union_set, int* flags_set, int nodeA, int nodeB, int flag) {
    get_parent(union_set, flags_set, nodeA);
    get_parent(union_set, flags_set, nodeB);
    if (union_set[nodeA] != union_set[nodeB]) {
        int parB = union_set[nodeB];
        union_set[parB] = union_set[nodeA];
        flags_set[parB] = (flag + flags_set[nodeA] + flags_set[nodeB]);
        return true;
    } else {
        return false;
    }
}


int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int t = 0;
    cin >> t;
    for (int _ = 0; _ < t; _++) {
        int n = 0;
        cin >> n;
        int** matrix = new int* [n];
        for (int i = 0; i < n; i++) {
            matrix[i] = new int [n];
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cin >> matrix[i][j];
            }
        }

        int current_type_num = n;
        int* union_set = new int [n];
        int* flags_set = new int [n];
        for (int i = 0; i < n; i++) {
            union_set[i] = i;
            flags_set[i] = 0;
        }

        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                if (matrix[i][j] < matrix[j][i]) {
                    bool flag = merge(union_set, flags_set, i, j, 0);
                    if (flag) {
                        current_type_num--;
                    }
                } else if (matrix[i][j] > matrix[j][i]) {
                    bool flag = merge(union_set, flags_set, i, j, 1);
                    if (flag) {
                        current_type_num--;
                    }
                }
                if (current_type_num == 1) {
                    break;
                }
            }
            if (current_type_num == 1) {
                break;
            }
        }

        for (int i = 0; i < n; i++) {
            get_parent(union_set, flags_set, i);
            if (flags_set[i] == 1) {
                for (int j = 0; j < n; j++) {
                    int temp = matrix[i][j];
                    matrix[i][j] = matrix[j][i];
                    matrix[j][i] = temp;
                }
            }
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cout << matrix[i][j] << " ";
            }
            cout << endl;
        }

        for (int i = 0; i < n; i++) {
            delete []matrix[i];
        }
        delete []matrix;
        delete []union_set;
        delete []flags_set;
    }
    return 0;
}